File: 윤서/자료구조/4949.py
# 4949 - 균형잡힌 문자열


import sys
input = sys.stdin.readline

sList = []

while True :                    # .이 입력될 때까지 실행
    string = input().rstrip()   # rstrip()이 없으면 .을 입력했을 때 반복문 종료되지 않음

    if string == '.' :          # 종료조건인 .이 입력되는 경우 while문 빠져나옴
        break
    stack = []

    for i in string:
        if i =='.':
            break
        elif i == '[' or i == '(':
            stack.append(i)
        elif i == ']':
            if len(stack) > 0 and stack[-1] == '[':
                stack.pop()
            else:
                stack.append(i)
                break
        elif i == ')':
            if len(stack) > 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)
                break

    if len(stack) == 0:
        print("yes")
    else:
        print("no")

File: 찬미/스택/4949.swift
while true {
    let input = readLine()!
    var flag = true
    // . 찍을 경우 문장 끝내기
    if input == "." {
        break
    }
    var stack: [String.Element] = []
    
    for i in input {
        // 괄호가 열렸을 경우 stack에 추가
        if i == "[" || i == "(" {
            stack.append(i)
        } else if i == "]" || i == ")" {
            // 이 밑에부터는 어떻게 코드를 짜면 좋을지 잘 모르겠네요 . . .
            // [ 와 ]만 짝이 될 수 있고, (와 )만 짝이 될 수 있는데
            // [이 나오고 )인 경우는 어떻게 구분을 해 줄 수 있을지?
            // 그 반대로 (이 나오고 ]이 나온 경우는…?
        }
    }
}
